boot/setefibootvars.go

```diff
@@ -0,0 +1,194 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"reflect"
+	"regexp"
+	"strconv"
+
+	"github.com/canonical/go-efilib"
+	"github.com/canonical/go-efilib/linux"
+)
+
+var (
+	ErrAllBootNumsUsed    = errors.New("all Boot#### variable numbers are already in use")
+	ErrNoMatchingVariable = errors.New("no variable matches the given boot option")
+	ErrInvalidBootOrder   = errors.New("BootOrder variable data must have even length")
+
+	defaultVarAttrs = efi.AttributeNonVolatile | efi.AttributeBootserviceAccess | efi.AttributeRuntimeAccess
+
+	efiListVariables = efi.ListVariables
+	efiReadVariable  = efi.ReadVariable
+	efiWriteVariable = efi.WriteVariable
+
+	linuxFilePathToDevicePath = linux.FilePathToDevicePath
+)
+
+// Converts the given efiFilePath to a device path, creates a load option
+// with that device path, and returns it in serialized form.
+func filePathToLoadOption(filePath string) ([]byte, error) {
+	devicePath, err := linuxFilePathToDevicePath(filePath, linux.ShortFormPathHD)
```

Comment:
I suppose you got it working. But trying to understand the code, I thought `filePath` was the relative path to the ESP. I am not sure how it can convert to a device path if it does not know what mount we are talking about. Are we missing `rootdir` from the grub bootloader object?

Reply:
You're right... From my previous discussions with Chris, I had misunderstood and was under the impression that the path should just be relative, but looking at the `go-efilib` code myself, it definitely needs `rootdir`. I've added this as part of a new `UefiBootloader` interface which handles creating the load option directly, instead of `makebootable.go` using trusted asset methods and then letting `setefibootvar.go` create the load option. The new way based on your idea is much more elegant I think. Open to any suggestions, of course!